Accurately count attribute values – 250 values, not chars

faisal-alvi opened the pull request:
### All Submissions:

<!-- Mark completed items with an [x] -->
* [x] Does your code follow the [WooCommerce Sniffs](https://github.com/woocommerce/woocommerce-sniffs/) variant of WordPress coding standards?
* [ ] Have you written new tests for your changes, as applicable?
* [ ] Have you successfully run tests with your changes locally?
* [ ] Will this change require new documentation or changes to existing documentation?

---

### Changes proposed in this Pull Request:
<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

**Fix attribute validation to count values instead of string length**

The WooCommerce Square plugin was incorrectly validating attribute values by checking the character length of pipe-separated strings instead of counting the actual number of attribute values. This caused validation failures for products with long attribute value names, even when they had fewer than 250 values, while allowing products with many short attribute values to pass validation even when exceeding the 250 limit.

This PR modifies `validateMaxAttributeValues()` function to count array values directly instead of converting to a string first

Closes https://linear.app/a8c/issue/SQUARE-147/attribute-limitation-is-250-values-but-plugin-js-checks-string-not.

### Steps to test the changes in this Pull Request:
<!-- Describe the steps to replicate the issue and confirm the fix -->
<!-- Try to include as many details as possible. -->

1. Create a product with one attribute containing 6 values, each with 50+ characters (e.g., "Very long attribute value name that exceeds typical length for testing purposes")
2. Press tab while editing; this will trigger a values check, confirm no errors are displayed. Previously, this would fail with "Too many attribute values" error.
3. Verify the product now syncs successfully since it has only 6 values (under the 250 limit)
4. Test with a product that has 251+ attribute values to confirm validation still works correctly and error is displayed (you can do this just by copy-pasting any string (for example "Red | ")
5. Verify error message

### Changelog entry

> Fix - Accurately count attribute values – 250 values, not characters.

faisal-alvi committed:
Accurately count attribute values – 250 values, not chars

dkotter commented on src/js/admin/wc-square-admin-products.js:
```diff
@@ -817,19 +817,23 @@ jQuery( document ).ready( ( $ ) => {
 				250,
 				$field,
-				wc_square_admin_products.i18n.too_many_attribute_values + ': ' + attrValues.substring(0, 50) + '...'
+				wc_square_admin_products.i18n.too_many_attribute_values + ': ' + valueCount + ' values (max 250)'
```
If we hardcode ` values (max 250)` here, this won't be translateable

dkotter commented on src/js/admin/wc-square-admin-products.js:
```diff
@@ -817,19 +817,23 @@ jQuery( document ).ready( ( $ ) => {
 		};
 
-		const validateMaxAttributeValues = ($field) => {
+	const validateMaxAttributeValues = ($field) => {
```
Do we have similar validation on the server-side? Or do we only validate this limit on the client-side? Just wondering if 1 - we need to apply this same fix to the server-side or 2 - add server-side validation if it doesn't exist

faisal-alvi replied:
We don't have server-side validation because Square's API handles the real enforcement. The client-side validation just provides immediate feedback to merchants on the product edit page.  We can keep the current approach (client-side validation only) unless there's a specific business need for server-side validation. What's your suggestion; should we add server-side validation when products are saved/updated, or is the current client-side + API failure approach sufficient?

faisal-alvi committed:
Remove hardcoded string and make it so its translateable

faisal-alvi committed:
Refine error message

faisal-alvi committed:
phpcs fix,
FILE: includes/Admin.php
-----------------------------------------------------------------------------------------------
FOUND 1 ERROR AND 0 WARNINGS AFFECTING 1 LINE
-----------------------------------------------------------------------------------------------
 163 | ERROR | A function call to __() with texts containing placeholders was found, but was not accompanied by a "translators:" comment on the line above to clarify the meaning of the placeholders. (WordPress.WP.I18n.MissingTranslatorsComment)
-----------------------------------------------------------------------------------------------

qasumitbagthariya commented:
### QA Update  ✅ 
-------------

I have verified this PR in the `fix/147-accurate-attribute-count-250-values` branch, which has been fixed and is functioning as intended.

I tested the following on this branch:
- Test with a product that has 251+ attribute values to confirm validation still works correctly and error is displayed  ✅ 

**Before**

<img width="1103" height="664" alt="image" src="https://github.com/user-attachments/assets/646373e5-5858-4999-9c31-aafdc69dfeca" />

**After**
<img width="1076" height="566" alt="image" src="https://github.com/user-attachments/assets/c6b60add-b635-47c7-9897-dbe1fcd1cae7" />


**Testing Environment** 

<Details>

- WordPress: 6.8.2
- Theme: Twenty Twenty-Five 1.2
- WooCommerce - 10.1.0
- PHP: 8.0.30
- Web Server: Nginx 1.20.2
- Browser: Chrome 
- OS: macOS 15.2 
- Branch: smoke-testing

</Details>


Steps to Test- As mentioned in the PR description.
Test Results - It is working as expected.
Functional Demo / Screencast - 
Special Notes -  Ready for UAT
Testing Document status:
Cases related to this Issue/PR are added to the Critical Flow Wiki pages: 

- [x] Yes
- [ ] Not Required/Applicable for this PR
------------------------------------------------	





@faisal-alvi @dkotter 

**Suggestion:**
It seems like most of the limitations are from the Square side, which we can handle in a similar way. For example:

1. If an attribute exceeds 65 characters, we should show an error at the time of entry itself, so that it doesn’t cause issues during sync.

faisal-alvi committed:
Merge branch 'trunk' into fix/147-accurate-attribute-count-250-values

qasumitbagthariya commented:
Regression / Smoke Test Report  ✅ 
-------

Tested with Archive File created via "php woorelease.phar build repo_URL" (Composer version 2.8.4, npm version 10.8.2, node version 20.18.1)

Status- Working expected with Plugin Archive/Zip file same as fix specific branch.

**Testing Environment** 

<Details>

- WordPress: 6.8.2
- Theme: Twenty Twenty-Five 1.3
- WooCommerce - 10.1.0
- PHP: 8.0.30
- Web Server: Nginx 1.20.2
- Browser: Chrome 
- OS: macOS 15.2 
- Branch: smoke-testing

</Details>


Next Step- Ready to Merge 🚀
